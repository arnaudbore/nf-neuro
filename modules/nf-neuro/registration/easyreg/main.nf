

process REGISTRATION_EASYREG {
    tag "$meta.id"
    label 'process_high'

    container "freesurfer/freesurfer:7.4.1"

    input:
    tuple val(meta), path(fixed_image), path(moving_image), path(fixed_segmentation), path(moving_segmentation)

    output:
    tuple val(meta), path("*_warped.nii.gz")                        , emit: image_warped
    tuple val(meta), path("*_warped_reference.nii.gz")              , emit: fixed_warped
    tuple val(meta), path("*_forward0_warp.nii.gz")                 , emit: forward_warp, optional: true
    tuple val(meta), path("*_backward0_warp.nii.gz")                , emit: backward_warp, optional: true
    tuple val(meta), path("*_warped_segmentation.nii.gz")           , emit: segmentation_warped, optional: true
    tuple val(meta), path("*_warped_reference_segmentation.nii.gz") , emit: fixed_segmentation_warped, optional: true
    path "versions.yml"                                             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def affine_only = task.ext.affine_only ? "--affine_only " : ""
    fixed_segmentation = "--ref_seg ${fixed_segmentation ?: "${prefix}_warped_segmentation.nii.gz" }"
    moving_segmentation = "--flo_seg ${moving_segmentation ?: "${prefix}_warped_reference_segmentation.nii.gz" }"
    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    mri_easyreg --ref $fixed_image \
        --flo $moving_image \
        --flo_reg ${prefix}_warped.nii.gz \
        --ref_reg ${prefix}_warped_reference.nii.gz \
        --fwd_field ${prefix}_forward0_warp.nii.gz \
        --bak_field ${prefix}_backward0_warp.nii.gz \
        $fixed_segmentation $moving_segmentation \
        --threads $task.cpus $affine_only

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        freesurfer: 7.4.1
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    mri_easyreg -h

    touch ${prefix}_warped.nii.gz
    touch ${prefix}_warped_reference.nii.gz
    touch ${prefix}_warped_segmentation.nii.gz
    touch ${prefix}_warped_reference_segmentation.nii.gz
    touch ${prefix}_forward0_warp.nii.gz
    touch ${prefix}_backward0_warp.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        freesurfer: 7.4.1
    END_VERSIONS
    """
}
