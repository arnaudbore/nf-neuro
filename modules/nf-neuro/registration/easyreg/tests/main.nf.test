nextflow_process {

    name "Test Process REGISTRATION_EASYREG"
    script "../main.nf"
    process "REGISTRATION_EASYREG"

    tag "modules"
    tag "modules_nfneuro"
    tag "registration"
    tag "registration/easyreg"

    tag "subworkflows"
    tag "subworkflows/load_test_data"


    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "freesurfer.zip", "registration.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("registration - easyreg") {
        config "./nextflow.config"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1: it.simpleName == "freesurfer"
                        b0: it.simpleName == "registration"
                    }
                ch_t1 = ch_split_test_data.t1.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1.nii.gz")
                    ]
                }
                ch_b0 = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]
                }
                input[0] = ch_t1
                    .join(ch_b0)
                    .map{ meta, t1, b0 -> [meta, t1, b0, [], []] }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.fixed_segmentation_warped.get(0).get(1)).name,
                    file(process.out.segmentation_warped.get(0).get(1)).name,
                    niftiMD5SUM(process.out.fixed_warped.get(0).get(1), 6),
                    niftiMD5SUM(process.out.image_warped.get(0).get(1), 6),
                    niftiMD5SUM(process.out.forward_warp.get(0).get(1), 6),
                    niftiMD5SUM(process.out.backward_warp.get(0).get(1), 6),
                    process.out.versions
                ).match() }
            )
        }
    }

    test("registration - easyreg - stub-run") {
        tag "stub"
        options "-stub-run"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1: it.simpleName == "freesurfer"
                        b0: it.simpleName == "registration"
                    }
                ch_t1 = ch_split_test_data.t1.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1.nii.gz")
                    ]
                }
                ch_b0 = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]
                }
                input[0] = ch_t1
                    .join(ch_b0)
                    .map{ meta, t1, b0 -> [meta, t1, b0, [], []] }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }
}
