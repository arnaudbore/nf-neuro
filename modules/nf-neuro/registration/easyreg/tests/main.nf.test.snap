{
    "registration - easyreg": {
        "content": [
            "test_warped_reference_segmentation.nii.gz",
            "test_warped_segmentation.nii.gz",
            "test_warped_reference.nii.gz:md5:header,c5e41f89848f91c53a9a7be44970d4b1,data,1501221fe23cd62bfdafb33367cadf4d",
            "test_warped.nii.gz:md5:header,ec5893cd9ea024e630c4444bd914c331,data,d75ae3fc935cd5cc70ea6797a4c70775",
            "test_forward0_warp.nii.gz:md5:header,0db0a80786ff39864cc17506ed1a0146,data,e4fa6c626729cbf236c34680e5c76df4",
            "test_backward0_warp.nii.gz:md5:header,74c92ee4cd3c4abfecf3da6cdb1d4650,data,8f26afa5c3de21469466213498542486",
            [
                "versions.yml:md5,4661210880c42a986923e8257f64c760"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-16T15:11:21.026914183"
    },
    "registration - easyreg - stub-run": {
        "content": [
            [
                "versions.yml:md5,4661210880c42a986923e8257f64c760"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-19T01:06:31.625578918"
    }
}